### ISSUE TITLE
Rule `consistent-return` does not behave properly inside async functions

### ISSUE THREAD

**Issue by marcospgp** [eslint/eslint#12865]
Inside an async function, calling
```js
return new Promise(resolve => {
  someAsyncSideEffect(err => {
    resolve();
  })
});
```
means the function returns an empty promise, just like what happens when simply `return;` is called.

But the `consistent-return` rule does not identify this correctly, and reports an error in the following code:

```js
async function example() {
  if (something) {
    someSyncSideEffect();
    return;
  } else {
    return new Promise(resolve => {
      someAsyncSideEffect(err => {
        resolve();
      }
    });
}
```

**Comment by mdjermanovic** [comment]
Thanks for the bug report!

I believe this is expected behavior:

> This rule requires return statements to either always or never specify values.

This is to warn on possible typing mistakes, which `return;` could be, as explained in the [documentation](https://eslint.org/docs/rules/consistent-return).

It's similar to how the rule reports an error in the following code:

```js
/*eslint consistent-return: ["error"]*/

function example() {
    if (something) {
        return;   
    } else {
        return undefined;       
    }
}
```

[Online Demo Link](https://eslint.org/demo#eyJ0ZXh0IjoiLyplc2xpbnQgY29uc2lzdGVudC1yZXR1cm46IFtcImVycm9yXCJdKi9cblxuZnVuY3Rpb24gZXhhbXBsZSgpIHtcbiAgICBpZiAoc29tZXRoaW5nKSB7XG4gICAgICAgIHJldHVybjsgICBcbiAgICB9IGVsc2Uge1xuICAgICAgICByZXR1cm4gdW5kZWZpbmVkOyAgICAgICBcbiAgICB9XG59Iiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjoxMSwic291cmNlVHlwZSI6InNjcmlwdCIsImVjbWFGZWF0dXJlcyI6e319LCJydWxlcyI6e30sImVudiI6e319fQ==)

Valid code for this rule would be to express the intention with `return undefined;` (or `void` something)

```js
/*eslint consistent-return: ["error"]*/

async function example() {
  if (something) {
    someSyncSideEffect();
    return undefined;
  } else {
    return new Promise(resolve => {
      someAsyncSideEffect(err => {
        resolve();
      });
    });
  }  
}
```

**Comment by marcospgp** [comment]
I see that the rule is not about what is technically being returned, and more about what the code looks like. Sounds good to me 👍

**Comment by darksabrefr** [comment]
Technically, the return type of an `async function` is always a `Promise` and never `undefined`, the rule documentation is false on this point. Moreover, in `async function`, the `return` statement does not mean you are returning a value. The two significations are:
- resolve (with or without a value) the `Promise` the `async function` returns
- delegate the `Promise` resolution to another `Promise`/`async function`. With classical `Promise` patterns the name of that is "chaining" (with `.then()` and/or `.catch()`).

That means using a `return` in those functions never means you are returning a value. The main issue is the rule doesn't do any difference between async procedures and functions (which are syntactically the same thing in javascript). For sync functions the rule considers as procedures those never using the `return` statement or using it but without return value (an stop function statement). All other are functions and must have a consistent `return`. This strategy is not applicable to `async function` without any adaptation because of the different `return` meaning.
Let's me show a simple example with this pseudo-code, this can be a base code for an auto-retrying function and it's an async procedure, not a function:
```javascript
async function fn () {
  const ok = await anotherFn();
  if (!ok) {
    return fn();
  }
}
```
In this example, the return type is consistent and implicit, it's a `Promise`. This `Promise` always resolves without any value. I have used `return` conditionally in its `Promise` delegation meaning. I've not forgotten any `return` statement. Adding `return undefined;` at the end is just an unnecessary and confusing thing to pass the rule, letting me think I resolve somewhere else with a value... but that's not the case !

Think too that `async function` is only syntactic sugar:
```javascript
async function demo () {
   ...
}
```
is strictly equivalent to
```javascript
function demo () {
  return new Promise((resolve, reject) => {
    ...
    resolve();
  })
}
```
In this second example, you can see that resolving a `Promise` in a classical way is always a procedure and not a function ! In fact, all `async function` are sync function (not procedure) wrapping an async function or an async procedure. This is the subtlety of this thing.

After this long argumentation, what is needed ? I think a minimal need is to have a option to disable the rule for `async function` because it fails to understand the meaning of this special functions type but can be useful for sync functions. A better solution would be the rule understands that `return aPromise();` inside an `async function` is not a `return` value statement of sync functions, and handles it like it does for the simple `return;` in both functions types.
